docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+}
+
+type LocalObjectTemplate struct {
+  // Ref is a required reference to a custom resource
+  // offered by a bootstrap provider.
+  Ref corev1.ObjectReference `json:"ref"`
+}
+```
+
+#### Modification to existing API Types
+##### Cluster
+1.  Add `Cluster.Spec.Managed` defined as
+    ```golang
+    // The information of the managed resources for the cluster.
+    // +optional
+    Managed *ManagedCluster `json:"managed,omitempty"`
+    ```
+1.  The `ManagedCluster` object has the following definition:
+    ```golang
+    // ManagedCluster encapsulates the information of the managed resources.
+    type ManagedCluster struct {
+      // The name of the ClusterClass object to create the managed cluster topology.
+      Class string `json:"class,omitempty"`
+
+      // The kubernetes version of the cluster.
+      Version string `json:"version"`
+
+      // RolloutAfter performs a rollout of the entire cluster one component at a time [...]
+      // +optional
+      RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`
+
+      // The information for the Control plane of the cluster.
+      ControlPlane ControlPlane `json:"controlPlane"`
+
+      // Worker encapsulates the different constructs that form the worker nodes
+      // for the cluster.
+      // +optional
+      Worker *Worker `json:"worker,omitempty"`
+    }
+    ```
+1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed cluster.
+    ```golang
+    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
+    type ControlPlane struct {
+     Metadata ObjectMeta `json:"metadata,omitempty"`
+
+     // The number of control plane nodes.
+     Replicas int `json:"replicas"`
+    }
+    ```
+1.  The `Worker` object represents the sets of worker nodes of the managed cluster.
+    ```golang
+    // Worker represents the different sets of worker nodes in the cluster.
+    type Worker struct {
+      Deployments []Deployment `json:”deployments,omitempty”`
+    }
+    ```
+1.  The `Deployment` object represents a single set of worker nodes of the managed cluster.
+    ```golang
+    // Deployment specifies the different parameters for a set of worker nodes in the cluster.
+    type Deployment struct {
+      Metadata ObjectMeta `json:"metadata,omitempty"`
+
+      // Class is the name of the DeploymentClass used to create the set of worker nodes.
+      // This should match one of the deployment classes defined in the ClusterClass object
+      // mentioned in the `Cluster.Spec.Class` field.
+      Class string `json:"class"`
+
+      // Name refers to the name of the set of worker nodes.
+      Name string `json:"name"`
+
+      // The number of worker nodes belonging to this set.
+      // If not specified, the cluster autoscaler for CAPI would be responsible
+      // for maintaining a min/max number of replicas.
+      // +optional
+      Replicas *int `json:"replicas,omitempty"`
+    }
```

Comment:
Where in this picture is the workerClass used?

Reply:
Got it, so https://github.com/kubernetes-sigs/cluster-api/pull/4678/files#r642478474

Reply:
WorkerClass is not really a "class", it's just there to make ClusterClass easier to extend in the future, if that should be necessary